### ISSUE TITLE
[geometry] Why does DOMQuad have readonly mutable points?

### ISSUE THREAD

**Issue by tabatkins** [w3c/fxtf-drafts#124]
_From @tabatkins on January 11, 2017 20:31_

DOMQuad's p1/p2/p3/p4 attributes are readonly, which means there's no setter.  However, the points themselves are DOMPoint, not DOMPointReadonly, so they're mutable.  This seems like a bizarre amalgam?  You *can* mutate the quad's points, you just can't reassign them.

We should either make DOMQuad fully mutable (removing the readonly annotations), or switch them to DOMPointReadonly, rename DOMQuad to DOMQuadReadonly, and if necessary, add a fully mutable DOMQuad as well.

_Copied from original issue: w3c/csswg-drafts#888_

**Comment by tabatkins** [comment]
_From @fantasai on March 30, 2017 16:0_

@tabatkins Can you refile this in the correct repo? Thanks~

**Comment by zcorpan** [comment]
Having readonly properties with mutable objects is quite common. 

For example
https://fetch.spec.whatwg.org/#request `headers`
https://html.spec.whatwg.org/#dom-document-head
https://html.spec.whatwg.org/#dom-dataset
https://dom.spec.whatwg.org/#dom-element-classlist
https://dom.spec.whatwg.org/#dom-element-attributes
https://drafts.csswg.org/css-font-loading/#dom-fontfacesource-fonts

...etc. So I think it doesn't follow that just because the object is mutable, the attribute must also be writable.

What is the use case for assigning the points in a `DOMQuad`?

I can think of a downside to changing them to be writable: code that assigns new points all the time (e.g. every animation frame) instead of assigning the primitive values on the points themselves would create a lot of garbage.

**Comment by zcorpan** [comment]
Another thing is that the API allows dictionaries to be passed in in general, but assigning a dictionary to an attribute wouldn't work I believe.
```
quad.p1 = {x:1, y:2};
```

**Comment by cvrebert** [comment]
See also https://lists.w3.org/Archives/Public/public-fx/2016OctDec/0016.html

**Comment by zcorpan** [comment]
I suppose we can add `DOMQuadReadOnly` that uses `DOMPointReadOnly` for `p1`..`p4`.

**Comment by tabatkins** [comment]
Rik offered the only pushback; otherwise it's just something that fell down the cracks. I'd prefer this be raised again in the group to attempt to gather interest; I still find the behavior quite weird.

(And I'm the person who raised this issue, not @smfr; they just raised a different issue that ended up with a similar conclusion that the current mutable/readonly mix is kinda weired.)

**Comment by tabatkins** [comment]
@zcorpan 

> Having readonly properties with mutable objects is quite common.

All of those examples are of things where producing the object in question *directly* would be quite rare, so the "readonly but mutable" weirdness doesn't matter in practice.

On the other hand, the geometry objects, particularly Points, are useful on their own, and it's reasonable to create those manually via their constructor.  It's weird that you can't change the points directly, and instead have to manually mutate their properties, when you can reasonable have Point objects that you'd like to use (or are pulling off of other objects).

**Comment by bzbarsky** [comment]
I haven't really been following the geometry API bits in Firefox....  I'm not sure who is right now, actually.  @smaug---- do you know?

**Comment by tabatkins** [comment]
> What do you think should happen when assigning a dictionary, though? (#124 (comment) above.) Or should there be dedicated methods to replace a point like quad.setP1({ x: 1, y: 2 }) ?

Assigning a dict would continue to throw a TypeError, as standard for something like this. If you have a dict, you should be able to just construct a DOMPoint from it and assign *that* - `quad.p1 = new DOMPoint({x:1, y:2})`.

**Comment by smaug----** [comment]
I don't know.

(DOMQuad looks a bit weird API anyhow. Why those static methods and not similar ctor? I would have made points use ReadOnly objects and getBounds() could have been readonly attribute bounds returning some ReadOnly object too)

**Comment by dirkschulze** [comment]
@smaug---- I am not sure why the overloading CTOR versions were replaced with the static version. Wasn’t that case initially. Though most implementations adapted the static method approach already. @zcorpan could you provide some info on that?

**Comment by domenic** [comment]
Yep. In general overloading, especially for constructors, is not something we want to spread.

**Comment by dirkschulze** [comment]
@zcorpan Thanks!

**Comment by mrbkap** [comment]
As an additional data point: currently, Chromium computes the `DOMQuad`'s bounds in the `DOMQuad` constructor, meaning that it's possible to construct an object whose bounds are out of sync with its current state. Is this a bug in the spec or Chromium?

**Comment by zcorpan** [comment]
I think that's a bug in chromium.

https://drafts.fxtf.org/geometry/ appears to be broken currently. @dirkschulze ?

**Comment by dirkschulze** [comment]
@zcorpan No idea what is up with the spec. BS file is present and looks good to me.

**Comment by dirkschulze** [comment]
@plinss Are you still managing the FX specs? Is there some service down?

**Comment by dirkschulze** [comment]
Actually, effecting all FX specs it seems?

**Comment by dirkschulze** [comment]
@tabatkins We have 3 fully interoperable implementations of `DOMQuad` with the specified behavior and did not get any additional push back on the API other than the above. Do you still request to have the API changed?

**Comment by tabatkins** [comment]
It's still weird and bad, yeah; my change wasn't rejected, the spec was just low on editor-hours for a while.

And the change I request (removing `readonly` from the attributes) should be backwards-compatible; all it'll do is possibly make some currently throwing code become valid, and we usually treat "errors stop being errors" as a pretty safe change to make, since people tend to rewrite code that throws errors.

**Comment by bzbarsky** [comment]
Setting a readonly property only throws in strict mode.  In non-strict mode it's a no-op.   So in terms of compat, it's a bit more complicated than going from throwing to non-throwing, unfortunately.

**Comment by tabatkins** [comment]
Ah dang, I thought it threw in all cases. :(

It would be unfortunate, but understandable, if this ended up unchangeable just due to compat weight built up in the interim time between me submitting the issue and an editor finally looking at it. However, I don't think it's reasonable to simply concede the issue as a result; Dirk, do you think you could look into this, then?
